### ISSUE TITLE
Neural Network for the training

### ISSUE THREAD

**Issue by MattiaMagro** [nndei/dVRL_Neri#4]
Hi @nndei 
I am trying to implement my scene in which I want to reach the target avoiding obstacles, but the success rate of the training, I perfomed, remains always 0. So, I was wondering if the increment in the number of layers of the neural network in _baselines.her.experiment.config.py_ could be useful.
In our scene, did you have to change the number of layers or the neurons in the layer to have success?

Thanks in advance, best regards

Mattia

**Comment by nndei** [comment]
Hello Mattia,

In deep learning, the number of layers is strictly correlated to the complexity of the function you want to learn, so I understand your concern about thinking that a more complex network could learn a more complex behaviour. 
However, it isn't that straight-forward in DRL. I wouldn't suggest tweaking anything in baselines yet. I suggest working with baselines' implementations and maybe switch to stable-baselines (https://github.com/hill-a/stable-baselines) sometimes later, as they offer a free hyperparameter optimiser called Optuna, which can be handy. 

What you should obtain from the standard implementations is a good success rate. Then, you may want to optimise (seek for a few % more) the hyperparameters for your specific task later on, using Optuna for example. 

Indeed, I used baselines' implementations of HER + DDPG and obtained a success_rate close to 100% after 800k time steps, but it highly depends on the number of parallel processes used, task and search space volume. 

You should follow this guideline: 
1. The **most important** and first step is to use a python script to run the task you want the robot to learn and check in the Docker container if the robot performs it smoothly. If it doesn't, then you're training something that doesn't even work when forced by a script.
2. If everything works when commanded by a script, try downloading and training a simpler task first. Perhaps, use the original scene and work in dVRL, which is far simpler https://github.com/ucsdarclab/dVRL.
3. Work your way through these factors. The success_rate highly depends on (A) The complexity of the task; (B) the search space volume (defined by where the robot can reach. If the robot could reach the whole R6 space, then a policy would be impossible to learn because there would be too much space to cover); (C) The number of processes you use: this is the number of parallel containers that run the training together. If you read the HER paper or the HER algorithm wiki provided by baselines and stable-baselines, or even the terminal logs when you run training using HER, you should read/know that HER performs best when using multiple parallel processes (19 in the HER paper were required to obtain the results they show). Therefore, it'll be hard to obtain a good success rate if you only run 1 process, no matter how much you wait; (D) whether you provide demonstrations or not: in the original dVRL paper, you can read that they employed demonstrations and used the Behavioural Cloning Loss, which can be turned on inside the hyperparameters file in baselines for the specific algorithms. In our work, demonstrations were mandatory to learn the pick task, in which the robot learns to pick an object. They weren't necessary to learn a policy for the reach task, in which the robot simply reaches an object. Demonstrations show the agent (the robot) how to perform the task, following the concept of Imitation Learning. dVRL's git should have an issue in which people asked the authors how they obtained their demonstrations. However, they shouldn't be necessary for a reach task.

I used 4 MPI processes (parallel processes aforementioned) to train my reach task. 
In this git, I also have a folder for Pick, but if you read the guide.txt, you should know that I tried working on it, but couldn't make it work before finishing my work on this project. 

Training using demonstrations on a simpler Pick task were collected and used by my supervisor and PhD candidate Claudia D'Ettorre from UCL. Her Git is https://github.com/Cladett

Regards,

Neri
